src/misc/msccl/msccl_status.cc

```diff
@@ -6,17 +6,66 @@
 #include "msccl/msccl_status.h"
 #include "msccl/msccl_struct.h"
 
-mscclStatus& mscclGetStatus() {
-  static mscclStatus status;
-  return status;
+#include "debug.h"
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+using namespace std;
+
+struct mscclRankState {
+  int rank;
+  bool initialized;
+  mscclStatus status;
+  mscclSavedProxyArgs savedProxyArgs;
+
+  mscclRankState() : rank(-1), initialized(false), status(), savedProxyArgs() {}
+  mscclRankState(const mscclRankState&) = delete;
+};
+
+static inline mscclRankState& mscclGetRankState(int rank) {
+  static mutex rankStatesMutex;
+  static unordered_map<int, shared_ptr<mscclRankState>> rankStates;
+
+  static thread_local shared_ptr<mscclRankState> threadRankState = make_shared<mscclRankState>();
+
+  if (rank < 0 || threadRankState->rank == rank) {
+    return *threadRankState;
+  }
+  if (threadRankState->rank >= 0) {
+    WARN("Changing rank %d to rank %d", threadRankState->rank, rank);
+  }
+
+  lock_guard<mutex> lock(rankStatesMutex);
+
+  auto rankStateIt = rankStates.find(rank);
+  if (rankStateIt == rankStates.end()) {
+    threadRankState->rank = rank;
+    rankStates.insert(make_pair(rank, threadRankState));
+  }
+  else {
+    threadRankState = rankStateIt->second;
+  }
+  return *threadRankState;
+}
+
+bool mscclInitialized(int rank) {
+  return mscclGetRankState(rank).initialized;
+}
+
+void mscclSetInitialized(int rank, bool initialized) {
+  mscclGetRankState(rank).initialized = initialized;
```

Comment:
Let's assert a failure here if it is initialized already. This way we can capture programming error or unsupported op mode

Reply:
Would you like an actual ASSERT or to log the error?

Reply:
I would like to see application to fail instantly with clear message with the reason. Otherwise we may get strange segfault or memory fault which takes really long time to figure out